File: packages/typescript-mdi-llmkit/src/jsonSurgery/placemarkedJSON.ts
/**
 * Converts an object to a JSON string, adding placemarks (comments) to indicate
 * the positions of objects and indexes for better readability.
 *
 * Example:
 * // root
 * {
 *   "name": "Movie Catalog",
 *   "description": "A catalog of movies",
 *   // root["items"]
 *   "items": [
 *     // root["items"][0]
 *     {
 *       "title": "Inception",
 *       "year": 2010,
 *       // root["items"][0]["keywords"]
 *       "keywords": [
 *         // root["items"][0]["keywords"][0]
 *         "sci-fi",
 *         // root["items"][0]["keywords"][1]
 *         "thriller"
 *       ]
 *     },
 *     // root["items"][1]
 *     {
 *       "title": "The Matrix",
 *       "year": 1999,
 *       // root["items"][1]["keywords"]
 *       "keywords": [
 *         // root["items"][1]["keywords"][0]
 *         "sci-fi",
 *         // root["items"][1]["keywords"][1]
 *         "action"
 *       ]
 *     }
 *   ],
 *   // root["catalog_metadata"]
 *   catalog_metadata: {
 *     "created_by": "Admin",
 *     "created_at": "2024-01-01"
 *   }
 * }
 *
 * @param obj The object to stringify
 * @param indent Number of spaces to use for indentation
 * @param skippedKeys Array of key names to skip/ignore in the output
 * @returns The placemarked JSON string
 */
export const placemarkedJSONStringify = (
  obj: any,
  indent?: number,
  skippedKeys?: string[]
): string => {
  indent = indent || 2;
  skippedKeys = skippedKeys || [];

  const indentStr = ' '.repeat(indent);
  const lines: string[] = [];

  function stringify(value: any, path: string, currentIndent: string): void {
    if (value === undefined || value === null) {
      lines.push(currentIndent + 'null');
    } else if (typeof value === 'boolean') {
      lines.push(currentIndent + value.toString());
    } else if (typeof value === 'number') {
      lines.push(currentIndent + value.toString());
    } else if (typeof value === 'string') {
      lines.push(currentIndent + JSON.stringify(value));
    } else if (Array.isArray(value)) {
      lines.push(`${currentIndent}[`);
      value.forEach((item, index) => {
        const itemPath = `${path}[${index}]`;
        // Add annotation for every array element
        lines.push(`${currentIndent}${indentStr}// ${itemPath}`);
        stringify(item, itemPath, currentIndent + indentStr);
        if (index < value.length - 1) {
          lines[lines.length - 1] += ',';
        }
        lines.push(``);
      });
      // Remove the final extra newline after the last array element if it exists
      if (lines[lines.length - 1] === '') {
        lines.pop();
      }
      lines.push(`${currentIndent}]`);
    } else if (typeof value === 'object') {
      lines.push(`${currentIndent}{`);
      const keys = Object.keys(value).filter(key => !(skippedKeys || []).includes(key));
      keys.forEach((key, index) => {
        const keyPath = `${path}["${key}"]`;
        const val = value[key];
        const isNonPrimitive = val !== null && typeof val === 'object';

        // Add annotation for non-primitive properties (objects and arrays)
        if (isNonPrimitive) {
          lines.push(``);
          lines.push(`${currentIndent}${indentStr}// ${keyPath}`);
        }

        lines.push(
          `${currentIndent}${indentStr}${JSON.stringify(key)}${isNonPrimitive ? ':' : ': '
          }`
        );
        stringify(val, keyPath, currentIndent + indentStr);
        if (index < keys.length - 1) {
          lines[lines.length - 1] += ',';
        }
      });
      lines.push(`${currentIndent}}`);
    } else {
      lines.push(`${JSON.stringify(value)}`);
    }
  }

  // Add root annotation
  lines.push('// root');
  stringify(obj, 'root', '');

  // Combine lines, handling the inline values properly
  let result = '';
  for (let i = 0; i < lines.length; i++) {
    const line = lines[i];
    if (line.endsWith(': ') && i + 1 < lines.length) {
      // This is a property key, combine with the next line if it's a simple value
      const nextLine = lines[i + 1];
      if (!nextLine.includes('{') && !nextLine.includes('[')) {
        result += line + nextLine.trim() + '\n';
        i++; // Skip the next line since we've already added it
        continue;
      }
    }
    result += line + '\n';
  }

  return result.trim();
};

/**
 * Navigates to a specific location in a JSON object based on the provided JSON path.
 * @param obj The JSON object to navigate
 * @param jsonPath An array representing the path to navigate (property names and array indexes)
 * @returns An object containing the parent of the target location, the key/index of the target,
 *          and the target object itself
 */
export const navigateToJSONPath = (obj: any, jsonPath: (string | number)[]): any => {
  // Navigate to the target location in the object
  let pathParent: any = null;
  let pathKeyOrIndex: string | number | null = null;
  let pathTarget: any = obj;
  for (let i = 0; i < jsonPath.length; i++) {
    pathKeyOrIndex = jsonPath[i] as string | number;

    pathParent = pathTarget;
    if (!pathParent) {
      throw new Error(
        `Error: Could not navigate to path ${JSON.stringify(jsonPath)};` +
        ` parent of path element ${JSON.stringify(pathKeyOrIndex)} is null or undefined.`
      );
    }
    pathTarget = pathParent[pathKeyOrIndex];
  }
  return { pathParent, pathKeyOrIndex, pathTarget };
};

File: packages/python-mdi-llmkit/src/mdi_llmkit/json_surgery/placemarked_json.py
"""Utilities for placemark-annotated JSON formatting and JSON-path navigation."""

import json

from typing import Any, Sequence


def placemarked_json_stringify(
    obj: Any,
    indent: int | None = None,
    skipped_keys: Sequence[str] | None = None,
) -> str:
    """Convert an object to JSON text with path placemark comments.

    The output mirrors the TypeScript implementation used by mdi-llmkit's
    `jsonSurgery` package, including comments such as `// root["items"][0]`.
    """
    indent = indent or 2
    skipped_keys = list(skipped_keys or [])

    indent_str = " " * indent
    lines: list[str] = []

    def _stringify(value: Any, path: str, current_indent: str) -> None:
        if value is None:
            lines.append(current_indent + "null")
        elif isinstance(value, bool):
            lines.append(current_indent + ("true" if value else "false"))
        elif isinstance(value, (int, float)):
            lines.append(current_indent + str(value))
        elif isinstance(value, str):
            lines.append(current_indent + json.dumps(value))
        elif isinstance(value, list):
            lines.append(f"{current_indent}[")
            for index, item in enumerate(value):
                item_path = f"{path}[{index}]"
                lines.append(f"{current_indent}{indent_str}// {item_path}")
                _stringify(item, item_path, current_indent + indent_str)
                if index < len(value) - 1:
                    lines[-1] += ","
                lines.append("")

            if lines and lines[-1] == "":
                lines.pop()
            lines.append(f"{current_indent}]")
        elif isinstance(value, dict):
            lines.append(f"{current_indent}{{")
            keys = [k for k in value.keys() if k not in skipped_keys]
            for index, key in enumerate(keys):
                key_path = f'{path}["{key}"]'
                val = value[key]
                is_non_primitive = val is not None and isinstance(val, (dict, list))

                if is_non_primitive:
                    lines.append("")
                    lines.append(f"{current_indent}{indent_str}// {key_path}")

                lines.append(
                    f"{current_indent}{indent_str}{json.dumps(key)}"
                    f"{':' if is_non_primitive else ': '}"
                )
                _stringify(val, key_path, current_indent + indent_str)
                if index < len(keys) - 1:
                    lines[-1] += ","
            lines.append(f"{current_indent}}}")
        else:
            lines.append(json.dumps(value))

    lines.append("// root")
    _stringify(obj, "root", "")

    result = ""
    i = 0
    while i < len(lines):
        line = lines[i]
        if line.endswith(": ") and i + 1 < len(lines):
            next_line = lines[i + 1]
            if "{" not in next_line and "[" not in next_line:
                result += line + next_line.strip() + "\n"
                i += 2
                continue

        result += line + "\n"
        i += 1

    return result.strip()


def navigate_to_json_path(obj: Any, json_path: Sequence[str | int]) -> dict[str, Any]:
    """Navigate to a location in a JSON-like object using a list path."""
    path_parent: Any = None
    path_key_or_index: str | int | None = None
    path_target: Any = obj

    for path_key_or_index in json_path:
        path_parent = path_target
        if not path_parent:
            raise ValueError(
                f"Error: Could not navigate to path {json.dumps(list(json_path))};"
                f" parent of path element {json.dumps(path_key_or_index)} is null or undefined."
            )
        path_target = path_parent[path_key_or_index]

    return {
        "path_parent": path_parent,
        "path_key_or_index": path_key_or_index,
        "path_target": path_target,
    }


placemarkedJSONStringify = placemarked_json_stringify
navigateToJSONPath = navigate_to_json_path
